Sources/MongoSwift/Operations/FindAndModifyOperation.swift

```diff
@@ -0,0 +1,163 @@
+
+        if mongocFlags != MONGOC_FIND_AND_MODIFY_NONE
+        && !mongoc_find_and_modify_opts_set_flags(self._options, mongocFlags) {
+            let remStr = String(describing: remove)
+            let upsStr = String(describing: upsert)
+            let retStr = String(describing: returnDocument)
+            throw UserError.invalidArgumentError(message:
+                "Error setting flags to \(flags); remove=\(remStr), upsert=\(upsStr), returnDocument=\(retStr)")
+        }
+
+        if let sort = sort {
+            guard mongoc_find_and_modify_opts_set_sort(self._options, sort.data) else {
+                throw UserError.invalidArgumentError(message: "Error setting sort to \(sort)")
+            }
+        }
+
+        // build an "extra" document of fields without their own setters
+        var extra = Document()
+        if let filters = arrayFilters { try extra.setValue(for: "arrayFilters", to: filters) }
+        if let coll = collation { try extra.setValue(for: "collation", to: coll) }
+
+        // note: mongoc_find_and_modify_opts_set_max_time_ms() takes in a
+        // uint32_t, but it should be a positive 64-bit integer, so we
+        // set maxTimeMS by directly appending it instead. see CDRIVER-1329
+        if let maxTime = maxTimeMS {
+            guard maxTime > 0 else {
+                throw UserError.invalidArgumentError(message: "maxTimeMS must be positive, but got value \(maxTime)")
+            }
+            try extra.setValue(for: "maxTimeMS", to: maxTime)
+        }
+
+        if let wc = writeConcern {
+            do {
+                try extra.setValue(for: "writeConcern", to: try BSONEncoder().encode(wc))
+            } catch {
+                throw RuntimeError.internalError(message: "Error encoding WriteConcern \(wc): \(error)")
+            }
+        }
+
+        guard extra.isEmpty || mongoc_find_and_modify_opts_append(self._options, extra.data) else {
+            throw UserError.invalidArgumentError(message: "Error appending extra fields \(extra)")
+        }
+    }
+
+    /// Sets the `update` value on a `mongoc_find_and_modify_opts_t`. We need to have this separate from the
+    /// initializer because its value comes from the API methods rather than their options types.
+    fileprivate func setUpdate(_ update: Document) throws {
+        guard mongoc_find_and_modify_opts_set_update(self._options, update.data) else {
+            throw UserError.invalidArgumentError(message: "Error setting update to \(update)")
+        }
+    }
+
+    fileprivate func setSession(_ session: ClientSession?) throws {
+        guard let session = session else {
+            return
+        }
+        var doc = Document()
+        try session.append(to: &doc)
+
+        guard mongoc_find_and_modify_opts_append(self._options, doc.data) else {
+            throw RuntimeError.internalError(message: "Couldn't read session information")
+        }
+    }
+
+    /// Cleans up internal state.
+    deinit {
```

Comment:
nit: prefer this at the top of the class declaration like a `defer` block, helps see that we _are_ indeed cleaning things up

Reply:
moved to the top.